### PR TITLE
Add support for NIFTI files with .nii and .nii.gz extensions.

### LINKED ISSUE #2972: NIFTI Image Reader
As requested below, implement a NIFTIImageReader in PyVista. 
vtk reader: https://vtk.org/doc/nightly/html/classvtkNIFTIImageReader.html

### Discussed in https://github.com/pyvista/pyvista/discussions/2971

<div type='discussions-op-text'>

<sup>Originally posted by **lucasgoddamn** July 11, 2022</sup>
Hey, im working on MRI Data and im quite new to programming and discovered pyvista for my purposes. 
Since you cant read NIFTI files with pv.read() i was wondering how i could transform my NIFTI files to a VTK or PLY and then plot it
Even though this might seem simple i haven't found any working solution yet, so i appreciate any quick help!</div>

### PR DESCRIPTION
**Pull request by jmargeta**
### Overview

NIFTI is a common format used in medical imaging (in particular for brain imaging), it extends the Analyze™ 7.5 format:
https://nifti.nimh.nih.gov/nifti-1
In the most common form, NIFTI filenames have the .nii extension or .hdr + .img when using the dual file format.
For compressed images, the convention is to use .nii.gz.

To read the files, we use the vtkNIFTIImageReader built into vtk. This commit therefore also extends the get_ext function to support the composed extensions with the .gz suffix.

The .hdr extension collides with the Radiance HDR file reader which is already handled by vtkHDRReader. A way around is to force_ext='.nii' with such files.

As discussed in: #2971
resolves #2972

### Details

- support to read NIFTI files (.nii and .nii.gz)
```python
image = pv.read('image.nii.gz')
```
- added `download_brain_atlas_with_sides` example which loads the `avg152T1_RL_nifti.nii.gz` NIFTI file from the [pyvista/vkt-data](https://github.com/pyvista/vtk-data/) repository. This is a test example which has the left - right orientation information embedded into the image itself:
```python
image = pv.examples.download_brain_atlas_with_sides()
image.slice(normal='z').plot(cpos='xy')
```
- image

- `get_ext` supports composite extensions (when the last one is .gz)

### CODE REVIEW

File: pyvista/examples/downloads.py
Comment (tkoyama010) [inline pyvista/examples/downloads.py:1743]:
```suggestion
    """Download an image of an averaged brain with a right-left label.
```

**Review by tkoyama010 — CHANGES_REQUESTED** [review]
Thanks for adding this. I guess we also need to add a test to `tests/utilities/test_reader.py` .

File: pyvista/examples/downloads.py
Comment (jmargeta) [reply pyvista/examples/downloads.py]: Thank you for your review @tkoyama010 ! I have just pushed the suggested changes.

Comment (jmargeta) [reply pyvista/examples/downloads.py]: I am not particularly set on how the function to retrieve the example NIfTI image should be named.
Does `download_brain_atlas_with_sides` make sense?

Comment (adeak) [reply pyvista/examples/downloads.py]: Sounds good to me.

Comment (jmargeta) [reply pyvista/examples/downloads.py]: Cool, thanks @adeak !

File: pyvista/utilities/reader.py
Comment (MatthewFlamm) [inline pyvista/utilities/reader.py:2047]: PyVista's convention is to end with a newline.

```suggestion

    """
```

**Review by MatthewFlamm — DISMISSED** [review]
This is great.  I only have one nitpick comment.

**Review by MatthewFlamm — APPROVED** [review]
LGTM. Great work!

**Comment by jmargeta** [comment]
Thanks @MatthewFlamm ! Your issue is a great reference to try out new vtk-based readers.

File: tests/utilities/test_reader.py
Comment (tkoyama010) [inline tests/utilities/test_reader.py:817]: Formatting was adapted to the previous and following functions.
```suggestion
    mesh = reader.read()
    assert all([mesh.n_points, mesh.n_cells])
```

**Review by tkoyama010 — APPROVED** [review]
LGTM. Thanks!

**Comment by jmargeta** [comment]
Thank you for all your reviews! :) Is there anything else from might end for this PR to be merge-able?

**Comment by akaszynski** [comment]
> Thank you for all your reviews! :) Is there anything else from my end for this PR to be merge-able?

Nope. We just like to wait 24 hours before merging to let all of our core developers have a chance to review.

**Comment by p-j-smith** [comment]
This is going to be super useful for me. I just have one suggestion (perhaps I should open an issue?). The VTK nifti reader ignores the origin and spacing specified in the nifti header file. It would be really useful if the correct origin and spacing were applied, something like this:

```python
import pyvista as pv
import pyvista.examples.downloads
import nibabel as nib
from vtk import vtkNIFTIImageReader

class NIFTIReader(pv.BaseReader):
    _class_reader = vtkNIFTIImageReader

image_filename = pv.examples.downloads._download_and_read('avg152T1_RL_nifti.nii.gz', load=False)
image = NIFTIReader(image_filename).read()
image_data = nib.load(image_filename)

origin = image_data.affine[:3, -1]
spacing = image_data.affine[:3, :3].diagonal()
image.SetOrigin(origin * np.sign(spacing))
image.SetSpacing(np.abs(spacing))
```

**Comment by akaszynski** [comment]
> This is going to be super useful for me. I just have one suggestion (perhaps I should open an issue?). The VTK nifti reader ignores the origin and spacing specified in the nifti header file. It would be really useful if the correct origin and spacing were applied, something like this:

At the risk of bloating this PR a bit, @jmargeta is this something you can add to the reader?

**Comment by MatthewFlamm** [comment]
This is something missing from the vtk reader it seems.  We may have different ideas about how to implement a workaround in PyVista.  For example, it could be an optional parameter in the read method.  Or it could be a setter/getter on the reader.  I propose a follow-on PR.

**Comment by jmargeta** [comment]
@p-j-smith, that is a great suggestion to have the origin and spacing fixed, thanks! For me (vtk 9.1.0), the spacing given by vtk (9.1.0) is already ok. It is only the origin that is set to (0, 0, 0). See experiments below.

@akaszynski, it is a question whether this PR shall depart from VTK's behavior.
I agree with @MatthewFlamm to handle this in a follow-on PR.

The zero origin is a [deliberate decision in VTK](https://gitlab.kitware.com/vtk/vtk/-/issues/17533) and is handled [differently from ITK for example](https://vtk.org/pipermail/vtk-developers/2016-November/034479.html).

Handling this should go hand in hand with how the oriented images and images with different axis ordering are treated in PyVista in general, and is linked to more image formats than just NIfTI (MetaImage).

# Experiments comparing different readers' outputs (different formats of the same image):
```python
import nibabel as nb
import numpy as np
import SimpleITK as sitk
import pyvista as pv

# Make an oriented image (inspired by real metadata) and save it as MetaImage and NIfTI
image_itk = sitk.Image([55, 55, 55], sitk.sitkInt16)
image_itk.SetOrigin((10.5, 21, 42))
image_itk.SetSpacing((1.25682, 1.36591, 1.39318))
image_itk.SetDirection((
    0.870293,
    -0.492534,
    3.92774e-07,
    0.46466,
    0.82104,
    0.331639,
    -0.163344,
    -0.288623,
    0.943406,
))
sitk.WriteImage(image_itk, 'image.mha', useCompression=True)
sitk.WriteImage(image_itk, 'image.nii.gz', useCompression=True)

# Read the images using PyVista (MetaImage and NIfTI) and Nibabel (NIfTI)
image_mha = pv.read('image.mha')
image_nii = pv.read('image.nii.gz')
image_nii_nb = nb.load('image.nii.gz')

# Get the spacing as described by p-j-smith 
spacing_nb = image_nii_nb.affine[:3, :3].diagonal().tolist()
origin_nb = image_nii_nb.affine[:3, -1] * np.sign(spacing_nb)

# Reading the matrices from vtkNIFTIImageReader to avoid nibabel
reader = vtk.vtkNIFTIImageReader()
reader.SetFileName('image.nii.gz')
reader.Update()
s_matrix = [[reader.GetSFormMatrix().GetElement(j, i) for i in range(4)] for j in range(4)]
q_matrix = [[reader.GetQFormMatrix().GetElement(j, i) for i in range(4)] for j in range(4)]


# Plot it all
print('Origin:')
print(f'itk:    {image_itk.GetOrigin()}')
print(f'pv-mha: {image_mha.origin}')
print(f'pv-nii: {image_nii.origin}')
print(f'nb-nii: {origin_nb}')

print('Spacing:')
print(f'itk:    {image_itk.GetSpacing()}')
print(f'pv-mha: {image_mha.GetSpacing()}')
print(f'pv-nii: {image_nii.GetSpacing()}')
print(f'nb-nii: {np.abs(spacing_nb).tolist()}')

print('Direction:')
print(f'itk:    {image_itk.GetDirection()}')
print(f'pv-mha: {image_mha.GetDirectionMatrix().GetData()}')
print(f'pv-nii: {image_nii.GetDirectionMatrix().GetData()}')
print(f'nb-nii_affine: \n{image_nii_nb.affine}')
print(f'nb-nii_sform: \n{image_nii_nb.get_sform()}')
print(f'nb-nii_qform: \n{image_nii_nb.get_qform()}')
with np.printoptions(precision=4):
   print(f'vtk-qmatrix: \n{np.array(q_matrix)}')
   print(f'vtk-smatrix: \n{np.array(s_matrix)}')

print('Physical dims:')
print(f'pv-mha: {image_mha.points.min()} - {image_mha.points.max()}')
print(f'pv-nii: {image_nii.points.min()} - {image_nii.points.max()}')
```

This is what I get

```
Origin:
itk:    (10.5, 21.0, 42.0)
pv-mha: (10.5, 21.0, 42.0)
pv-nii: (0.0, 0.0, 0.0)
nb-nii: [10.5 21.  42. ]
Spacing:
itk:    (1.25682, 1.36591, 1.39318)
pv-mha: (1.25682, 1.36591, 1.39318)
pv-nii: (1.2568199634552002, 1.3659100532531738, 1.393180012702942)
nb-nii: [1.0938016176223755, 1.121466875076294, 1.3143348693847656]

Direction:
itk:    (0.870293, -0.492534, 3.92774e-07, 0.46466, 0.82104, 0.331639, -0.163344, -0.288623, 0.943406)
pv-mha: (1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0)
pv-nii: (1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0)
nb-nii_affine:
[[-1.09380162e+00  6.72757506e-01 -5.42277462e-07 -1.05000000e+01]
 [-5.83993733e-01 -1.12146688e+00 -4.62032974e-01 -2.10000000e+01]
 [-2.05293924e-01 -3.94233108e-01  1.31433487e+00  4.20000000e+01]
 [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  1.00000000e+00]]
nb-nii_sform:
[[-1.09380162e+00  6.72757506e-01 -5.42277462e-07 -1.05000000e+01]
 [-5.83993733e-01 -1.12146688e+00 -4.62032974e-01 -2.10000000e+01]
 [-2.05293924e-01 -3.94233108e-01  1.31433487e+00  4.20000000e+01]
 [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  1.00000000e+00]]
nb-nii_qform:
[[-1.09380170e+00  6.72757471e-01 -5.47204230e-07 -1.05000000e+01]
 [-5.83993723e-01 -1.12146686e+00 -4.62032964e-01 -2.10000000e+01]
 [-2.05293917e-01 -3.94233092e-01  1.31433485e+00  4.20000000e+01]
 [ 0.00000000e+00  0.00000000e+00  0.00000000e+00  1.00000000e+00]]
vtk-qmatrix:
[[-8.7029e-01  4.9253e-01 -3.9277e-07 -1.0500e+01]
 [-4.6466e-01 -8.2104e-01 -3.3164e-01 -2.1000e+01]
 [-1.6334e-01 -2.8862e-01  9.4341e-01  4.2000e+01]
 [ 0.0000e+00  0.0000e+00  0.0000e+00  1.0000e+00]]
vtk-smatrix:
[[-8.7029e-01  4.9253e-01 -3.8924e-07 -1.0500e+01]
 [-4.6466e-01 -8.2104e-01 -3.3164e-01 -2.1000e+01]
 [-1.6334e-01 -2.8862e-01  9.4341e-01  4.2000e+01]
 [ 0.0000e+00  0.0000e+00  0.0000e+00  1.0000e+00]]

Physical dims:
pv-mha: 10.5 - 117.23172000000004
pv-nii: 0.0 - 75.23172068595886
```

Note, that for oriented/non-axis-aligned images or images with flipped axes the above mention method using diagonal of the affine matrix cannot be used to obtain spacing and origin.

Getting the origin without nibabel should be possible from the last column of the s and q matrices (with extra caution to get the signs right) - see vtk-smatrix, vtk-qmatrix.

**Comment by MatthewFlamm** [comment]
I'm not familiar at all with this data type.   It seems like exposing `reader.GetSFormMatrix()` and similar as properties in PyVista would be useful as a first step.  Up to you whether to include in this PR or a follow-on PR.

**Comment by jmargeta** [comment]
Is the addition of properties to the reader like below what you have in mind @MatthewFlamm ? 

```python
class NIFTIReader(BaseReader):
    """NIFTI Reader for .nii and .nii.gz files.

    Examples
    --------
    >>> import pyvista
    >>> from pyvista import examples
    >>> filename = examples.download_brain_atlas_with_sides(load=False)
    >>> filename.split("/")[-1]  # omit the path
    'avg152T1_RL_nifti.nii.gz'
    >>> reader = pyvista.get_reader(filename)
    >>> mesh = reader.read()
    >>> mesh.plot()

    """

    _class_reader = _vtk.vtkNIFTIImageReader

    @property
    def qform_matrix(self):
        return array_from_vtkmatrix(self.reader.GetQFormMatrix())
    
    @property
    def qform_code(self):
        return self._header.GetQFormCode()

    @property
    def sform_matrix(self):
        return array_from_vtkmatrix(self.reader.GetSFormMatrix())

    @property
    def sform_code(self):
        return self._header.GetSFormCode()
    
    @property
    def qfac(self):
        return self.reader.GetQFac()

    @property
    def _header(self):
        return self.reader.GetNIFTIHeader()



reader = pyvista.NIFTIReader(filename)
mesh = reader.read()
assert reader.sform_matrix.shape == (4, 4)
```

Beyond that, I am currently uncertain on how to integrate the orientation info into the loaded `pv.UniformGrid` instance.
That is why I would prefer to go with how VTK handles the loading for now and to have a follow-up PR later.

**Comment by MatthewFlamm** [comment]
> Is the addition of properties to the reader like below what you have in mind @MatthewFlamm ?

Exactly what I had in mind.



> Beyond that, I am currently uncertain on how to integrate the orientation info into the loaded `pv.UniformGrid` instance.
> That is why I would prefer to go with how VTK handles the loading for now and to have a follow-up PR later.

I think this is a good approach.

**Comment by jmargeta** [comment]
Super! Then, I will keep this for the next PR.

## For future reference:
 - sform is the matrix coming from DICOM
 - qform is the approximate - orthogonalized (no skew) version of sform
 - In most of the cases, when the sform_matrix is present, the origin could be computed from the more precise qform as:
`reader.sform_matrix[:3, 3] * np.array([-1, -1, 1])`, otherwise qform should be used instead.
 - there are [quite a few](https://github.com/InsightSoftwareConsortium/ITK/blob/c1839700a9d28946d1313caf91795e4ce160eb13/Modules/IO/NIFTI/src/itkNiftiImageIO.cxx#L1805) things to consider - doing this in NumPy version would be way simpler
 - the initial thing could look something like this:

```python
class NIFTIReader(BaseReader):
    """NIFTI Reader for .nii and .nii.gz files.

    Examples
    --------
    >>> import pyvista
    >>> from pyvista import examples
    >>> filename = examples.download_brain_atlas_with_sides(load=False)
    >>> filename.split("/")[-1]  # omit the path
    'avg152T1_RL_nifti.nii.gz'
    >>> reader = pyvista.get_reader(filename)
    >>> mesh = reader.read()
    >>> origin = reader.origin
    >>> mesh.plot()

    """

    _class_reader = _vtk.vtkNIFTIImageReader

    @property
    def sform_matrix(self):
        if self.reader.GetSFormMatrix() is not None:
            return array_from_vtkmatrix(self.reader.GetSFormMatrix())

    @property
    def qform_matrix(self):
        if self.reader.GetQFormMatrix() is not None:
            mat = array_from_vtkmatrix(self.reader.GetQFormMatrix())
            has_no_skew = np.allclose(mat[:3, :3] @ mat[:3, :3].T, np.eye(3))
            if not has_no_skew:
                logging.warning("The qform matrix should have no skew")
            return mat

    @property
    def sform_code(self):
        return self._header.GetSFormCode()

    @property
    def qform_code(self):
        return self._header.GetQFormCode()

    @property
    def qfac(self):
        return self.reader.GetQFac()

    @property
    def origin(self):
        if self._is_old_analyze_format:
            logging.warning("Using old Analyze format and the origin is set to (0, 0, 0)")
            return np.array([0, 0, 0])
        if self.sform_matrix is not None:
            return self.sform_matrix[:3, 3] * np.array([-1, -1, 1])
        if self.qform_matrix is not None:
            return self.qform_matrix[:3, 3] * np.array([-1, -1, 1])
        logging.warning("Direction matrices are not found, the origin is set to (0, 0, 0)")
        return np.array([0, 0, 0])

    @property
    def _is_old_analyze_format(self):
        return self.qform_code == 0 and self.sform_code == 0

    @property
    def _header(self):
        return self.reader.GetNIFTIHeader()

```

**Comment by jmargeta** [comment]
Splendid, thank you all for your help with this PR!

**Pull request merged**
